easybuild/tools/containers.py

```diff
@@ -0,0 +1,304 @@
+
+    # name of Singularity definition file
+    img_name = build_option('container_image_name')
+    if img_name:
+        def_file_label = os.path.splitext(img_name)[0]
+    else:
+        def_file_label = mod_names[0].replace('/', '-')
+
+    def_file = 'Singularity.%s' % def_file_label
+
+    # adding all the regions for writing the  Singularity definition file
+    content = SINGULARITY_TEMPLATE % {
+        'bootstrap': bootstrap_agent,
+        'from': bootstrap_from,
+        'install_os_deps': install_os_deps,
+        'easyconfigs': ' '.join(os.path.basename(e['spec']) for e in easyconfigs),
+        'mod_names': ' '.join(mod_names),
+    }
+    def_path = os.path.join(cont_path, def_file)
+
+    if os.path.exists(def_path):
+        if build_option('force'):
+            print_msg("WARNING: overwriting existing container recipe at %s due to --force" % def_path)
+        else:
+            raise EasyBuildError("Container recipe at %s already exists, not overwriting it without --force", def_path)
+
+    write_file(def_path, content)
+    print_msg("Singularity definition file created at %s" % def_path, log=_log)
+
+    return def_path
+
+
+def build_singularity_image(def_path):
+    """Build Singularity container image by calling out to 'singularity' (requires admin privileges!)."""
+
+    cont_path = container_path()
+    def_file = os.path.basename(def_path)
+
+    # use --imagename if specified, otherwise derive based on filename of recipe
+    img_name = build_option('container_image_name')
+    if img_name is None:
+        # definition file Singularity.<app>-<version, container name <app>-<version>.<img|simg>
+        dot_idx = def_file.find('.')
+        img_name = def_file[dot_idx+1:]
+
+    cmd_opts = ''
+
+    image_format = build_option('container_image_format')
+
+    # squashfs image format (default for Singularity)
+    if image_format in [None, CONT_IMAGE_FORMAT_SQUASHFS]:
+        img_path = os.path.join(cont_path, img_name + '.simg')
+
+    # ext3 image format, creating as writable container
+    elif image_format == CONT_IMAGE_FORMAT_EXT3:
+        img_path = os.path.join(cont_path, img_name + '.img')
+        cmd_opts = '--writeable'
+
+    # sandbox image format, creates as a directory but acts like a container
+    elif image_format == CONT_IMAGE_FORMAT_SANDBOX:
+        img_path = os.path.join(cont_path, img_name)
+        cmd_opts = '--sandbox'
+
+    else:
+        raise EasyBuildError("Unknown container image format specified for Singularity: %s" % image_format)
+
+    if os.path.exists(img_path):
+        if build_option('force'):
+            print_msg("WARNING: overwriting existing container image at %s due to --force" % img_path)
+        else:
+            raise EasyBuildError("Container image already exists at %s, not overwriting it without --force", img_path)
+    else:
+        cmd = "sudo singularity build %s %s %s" % (cmd_opts, img_path, def_path)
```

Comment:
Maybe you should check for sudo existence while checking for singularity image build tools presence.

Reply:
@FooBarQuaxx Makes sense, I'll check for `sudo` too.

@shahzebsiddiqui I'm not seeing that issue (on `macOS`, with a fake `singularity`), I get a proper password prompt, which repeats if I mistype the password...

```
== Running 'sudo singularity build  /Users/kehoste/.local/easybuild/containers/cURL-7.58.0-GCCcore-6.4.0.simg /Users/kehoste/.local/easybuild/containers/Singularity.cURL-7.58.0-GCCcore-6.4.0', you may need to enter your 'sudo' password...
Password:
Password:
== Singularity image created at /Users/kehoste/.local/easybuild/containers/cURL-7.58.0-GCCcore-6.4.0.simg
== Temporary log file(s) /tmp/eb-5596mm/easybuild-L8YGho.log* have been removed.
== Temporary directory /tmp/eb-5596mm has been removed.
```

Can you try again with `eb -ld` to see where it hangs?

Reply:
@shahzebsiddiqui Note that it's possible that `sudo` doesn't ask your password if you've recently entered it, so if you see a 'hang' it's probably because `singularity` is running...